packager/amp_cache_transform/amp_cache_transform.go

```diff
@@ -0,0 +1,139 @@
+// Parses and interprets the AMP-Cache-Transform header specified at
+// https://github.com/ampproject/amphtml/blob/master/spec/amp-cache-transform.md.
+
+package amp_cache_transform
+
+import (
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-3.3
+type parameterisedIdentifier struct {
+	id     string
+	// In the future, this will include params.
+}
+
+// https://tools.ietf.org/html/rfc7230#appendix-B (OWS = "optional whitespace")
+func discardOWS(reader *strings.Reader) error {
+	for reader.Len() > 0 {
+		c, err := reader.ReadByte()
+		if err != nil {
+			return errors.Wrap(err, "reading OWS")
+		}
+		if c != ' ' && c != '\t' {
+			// Return to the position before the non-whitespace char.
+			reader.Seek(-1, io.SeekCurrent)
+			return nil
+		}
+	}
+	return nil
+}
+
+// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-4.2.3
+func parseParameterisedList(reader *strings.Reader) ([]parameterisedIdentifier, error) {
+	items := []parameterisedIdentifier{}
+	for reader.Len() > 0 {
+		item, err := parseParameterisedIdentifier(reader)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing parameterised identifier")
+		}
+		items = append(items, *item)
+		// Steps 3 and 4 are swapped per https://github.com/httpwg/http-extensions/pull/667.
+		if reader.Len() == 0 {
+			return items, nil
+		}
+		if err := discardOWS(reader); err != nil {
+			return nil, errors.Wrap(err, "discarding OWS")
+		}
+		comma, err := reader.ReadByte()
+		if err != nil {
+			return nil, errors.Wrap(err, "reading comma")
+		}
+		if comma != ',' {
+			return nil, errors.New("expected comma")
+		}
+		if err := discardOWS(reader); err != nil {
+			return nil, errors.Wrap(err, "discarding OWS")
+		}
+		if reader.Len() == 0 {
+			return nil, errors.New("expected another param-id")
+		}
+	}
+	return nil, errors.New("reached unexpected end of parseParameterisedList")
```

Comment:
nit. The only way you can exit the for loop without returning is if the input reader is empty to begin with, right?

There are no break statements, and the end of the for loop body return exits if the for loop condition would fail.

It seems that this error message could be more specific ("empty parameterized list") or you could eliminate the last "len() = 0" check at the end of the for body and make this generic.

Reply:
Done. I tried to copy the spec'd algorithm as closely as possible, and for the life of me, couldn't figure out what this error was even doing! Case of the Fridays. I'd mistakenly thought it was an "unreachable" kind of thing.